src/Functions/h3PolygonToCells.cpp

```diff
@@ -0,0 +1,187 @@
+
+                // will be ring, polygon, or multipolygon (boost::geometry)
+                auto geometry = geometries[row];
+                boost::geometry::correct(geometry); // TODO: Is this good or bad?
+
+                // always work on multipolygon
+                CartesianMultiPolygon multi_polygon;
+                if constexpr (std::is_same_v<ColumnToMultiPolygonsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = geometry;
+                else if constexpr (std::is_same_v<ColumnToPolygonsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = boost::geometry::model::multi_polygon({ geometry });
+                else if constexpr (std::is_same_v<ColumnToRingsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = boost::geometry::model::multi_polygon({ boost::geometry::model::polygon({ geometry }) });
+
+                // will only iterate once for ring and polygon
+                for (auto & polygon : multi_polygon)
+                {
+                    GeoPolygon h3_polygon{};
+                    parseGeoLoopFromRing(polygon.outer(), &h3_polygon.geoloop);
+
+                    auto holes = polygon.inners();
+                    int count = static_cast<int>(std::size(holes));
+                    h3_polygon.holes = static_cast<GeoLoop*>(malloc(count * sizeof(GeoLoop)));
+                    h3_polygon.numHoles = count;
+                    int i = 0;
+                    for (auto & hole : holes)
+                    {
+                        parseGeoLoopFromRing(hole, &h3_polygon.holes[i++]);
+                    }
+
+                    const size_t vec_size = maxPolygonToCellsSize(&h3_polygon, resolution);
+                    if (vec_size > MAX_ARRAY_SIZE)
+                        throw Exception(
+                            ErrorCodes::TOO_LARGE_ARRAY_SIZE,
+                            "The result of function {} (array of {} elements) will be too large with resolution argument = {}",
+                            getName(), vec_size, toString(resolution));
+
+                    std::vector<H3Index> hindex_vec;
+                    hindex_vec.resize(vec_size);
+                    polygonToCells(&h3_polygon, resolution, hindex_vec.data());
+
+                    dst_data.reserve(dst_data.size() + vec_size);
+                    for (auto hindex : hindex_vec)
+                    {
+                        if (hindex != 0)
+                        {
+                            ++current_offset;
+                            dst_data.insert(hindex);
+                        }
+                    }
+
+                    dst_offsets[row] = current_offset;
+                }
+            }
+        }
+        );
+
+
+        return dst;
+    }
+
+private:
+    /** Converts from boost ring to h3 loop
+    * Can be used for both polygon exterior and holes
+    */
+    void parseGeoLoopFromRing(auto ring, GeoLoop* geoloop) const
```

Comment:
It feels like this method could be a separate SQL function, no?

Reply:
The function builds an H3-specific geometry struct. I didn't think it made sense in a SQL context. What should the SQL function return, if I wanted to pursue this?